### DISCUSSION TITLE
Media type registration

### DISCUSSION THREAD

**Discussion by jdesrosiers** [json-schema-org/community#198] [Specification]
The JSON Schema specification defines the `application/schema+json` and `application/schema-instance+json`, but these media types have never been officially registered with the IANA. We're in the process of getting those media types registered. We met yesterday to discuss our approach and some potential problems.

What we define should work for all past a future releases of JSON Schema and their existing implementations as well as custom versions such as OpenAPI and MongoDB. In other words, we're not designing something ideal from scratch. We're defining something that is compatible with how past and current implementation use these media types.

The approach we are taking is to standardize only dialect identification in the media type registration and deferring the rest of how to interpret the document to the specific JSON Schema specification that was identified. For example, if `http://json-schema.org/draft-07/schema#` is identified, then the document should be interpreted based on the draft-07 specification.

These are all ways the dialect can be identified with items at the top of the list taking precedence over items lower in the list.
> * The `$schema` keyword
> * The "schema" media type parameter
> * The context of the enclosing document. This applies only when a schema is embedded within a document. The enclosing document could be another schema in the case of a bundled schema or it could be another type of document that includes schemas such as an OpenAPI document.
> * If none of the above result in identifying the dialect, client behavior is undefined.

However, this doesn't technically work for 2019-09 and 2020-12 because it's not `$schema` that determines how a schema is processed, it's `$vocabulary`. That leaves us with only bad options. Here are some of the things we discussed.

1. Standardize `$vocabulary` as well as `$schema`.
    This isn't a good option because `$vocabulary` is still evolving and is likely to change.
2. Introduce a new keyword that declares the core vocabulary and standardize that only instead of all of `$vocabulary`.
    That's more likely to be stable than `$vocabulary`, but since it's a new keyword and part of the vocabulary system that is likely to change, it's very possible we don't get this right the first time.
3. Use the `$schema` of the meta-schema instead of the core vocabulary declared in `$vocabulary` to determine how to process the schema.
    Although technically not correct, this seems to be the approach that most implementations have taken because using `$vocabulary` for core vocabulary identification is somewhat problematic. There were some comments that this approach would allow you do some strange things. I didn't follow that in the discussion. This would be a good place to give an example of what that meant.

**Comment by jdesrosiers** [comment]
Here's another option. It's not perfect either, but while writing this up I remembered that this was how I reconciled this problem in the first place. If this is how we want to approach it, the registration should be updated to make this clear.

It should say that if the identified dialect is not directly understood by the client, information about how to process the schema may be encoded in the meta-schema identified by the dialect id. Then implementations that support 2019-09 and 2020-12 know to look for a known core vocabulary declaration in the `$vocabulary` keyword.

That's pretty open ended, but it's compatible with all existing versions and doesn't constrain us in the future. We're free to change `$vocabulary` however we want to in the future or use some other mechanism entirely.

It's not perfect because it doesn't fully define how to determine which specification to use to process the schema, it delegates some of that the identification process to individual dialects. But, it works and it's compatible with past and future. I think this is the best option we have.

**Reply by handrews** [reply]
How is this option different from the option of (at least partially) standardizing `$vocabulary`?  We can't just say "look in `$vocabulary` if `$vocabulary` might change, because you won't know which version of `$vocabulary` you have until after looking at it and finding the core vocabulary.  Or am I missing something here?

**Reply by jdesrosiers** [reply]
Yes, it's a bit of a chicken and the egg problem, but I realized that there is a way to make it work. If your client supports 2020-12 and you want to check if it's a 2020-12 schema, you can start by assuming that it's a 2020-12 schema. Given that assumption, the 2020-12 specification says that it should have the 2020-12 core vocabulary declaration the `$vocabulary` keyword. If the proper declaration is found, your assumption was correct and you are working with a 2020-12 schema. If any part of that vocabulary check fails, your assumption was wrong and you know this is not a 2020-12 schema.

```
IF $schema is a known dialect
  schemaType = $schema
ELSE IF the meta-schema has $vocabulary["https://json-schema.org/draft/2019-09/vocab/core"] == true
  schemaType = "https://json-schema.org/draft/2019-09/vocab/core"
ELSE IF the meta-schema has $vocabulary["https://json-schema.org/draft/2020-12/vocab/core"] == true
  schemaType = "https://json-schema.org/draft/2020-12/vocab/core"
ELSE
  I don't know how to process this schema. I might guess, use a default, or throw an error.
```

**Reply by jdesrosiers** [reply]
It just occurred to me that given this process, we really don't even need to standardize `$schema`. That can be delegated to individual specifications as well.

**Reply by karenetheridge** [reply]
> However, this doesn't technically work for 2019-09 and 2020-12 because it's not $schema that determines how a schema is processed, it's $vocabulary.

I don't see how this follows. `$vocabulary` doesn't need to be standardized at this level, because it's defined by the draft2020-12 specification.  For example, say draft2022-07 is released and we rename the `$vocabulary` keyword to `$furble` -- we still don't need to define `$furble` or `$vocabulary` at the media type level, because the significance of those keywords is only determined after we have finished parsing `$schema` and know what version of the spec we are dealing with. As long as the implementation knows the draft2022-07 specification version, it will know what to do with `$furble` and can parse the remaining keywords in the document. If it doesn't, it can't, and it can error right away.

Only `$schema` (and possibly `$id`) need to remain constant across specification versions i.e. at the media type level.

edit: this is still true if using a custom metaschema - e.g.`"$schema": "https://example.com/my-meta-schema"`. We just need to describe the semantics for looking in the document identified by "https://example.com/my-meta-schema" and examining _its_ $schema keyword, and so on in sequence until we finally either reach a $schema that is a URI we have already looked at (that is, an `$id` of a document we have already examined, and we are in a circular loop with no resolution) or it is a specification metaschema that we know about (e.g. draft2020-12).   I'm pretty sure I've described this process before in previous threads :)

**Reply by awwright** [reply]
> I don't see how this follows. 

`$schema` doesn't correlate to a draft. It's a heuristic that works if the schema uses only standard keywords, but advanced usage just won't work predictably. For example:

> We just need to describe the semantics for looking in the document identified by "https://example.com/my-meta-schema" and examining its $schema keyword

This doesn't work because you can write a meta-schema referring to the 2019-09 meta-schema but that uses 2020-12 vocabularies.

You might even write your own meta-schema that refers to itself (like the hyper-schema does), and the only hint that you're using 2020-12 is the fact it refers to 2020-12 in $vocabulary.

**Reply by karenetheridge** [reply]
My point is that everything can be derived from `$schema` and then parsing the document it refers to (and so on and so on), so this is the only thing that needs to be fixed at the media-type level.

**Reply by jdesrosiers** [reply]
@karenetheridge That's what I was describing in Option 3 above. It's what you, me, and Greg all independently converged on in our implementations, but it's not technically correct for the 2019-09 and 2020-12 specifications.

It's not `"$schema": "https://json-schema.org/draft/2020-12/schema"` that identifies the schema as a 2020-12 schema, it's `"$vocabulary": { "https://json-schema.org/draft/2020-12/vocab/core": true, ... }`.

**Reply by karenetheridge** [reply]
In a separate thread I want to pursue the concept of mixing vocabularies from different spec versions and what that might look like in a few examples, because I don't think it's very clear (and maybe we want to outright prohibit it, if it gets too complicated - and then the whole mess gets a little clearer).

**Reply by karenetheridge** [reply]
> It's not "$schema": "https://json-schema.org/draft/2020-12/schema" that identifies the schema as a 2020-12 schema ...

This is not entirely true. The `$vocabulary` keywords in the metaschema determine which keywords are to be used in the evaluation (and how those keywords behave), but there is more to the spec than that - for example the overall semantics of evaluation, URI resolution etc.

I guess it comes down to how much of the behaviour is determined purely from the keywords themselves, and how much is of broader scope than that.  Do we need to come up with some pathological examples of mixing and matching $vocabulary and $schema values to see how weird it can get? I almost feel like this is an academic distinction that isn't going to have much impact in the practical sense. For example: what is the *actual difference in behaviour* between this:

```
$schema: https://json-schema.org/draft/2019-09/schema
$vocabulary:
  https://json-schema.org/draft/2020-12/vocab/core: true
  .. all the rest of the standard 2020-12 vocabularies...
```

and this:

```
$schema: https://json-schema.org/draft/2020-12/schema
$vocabulary:
  https://json-schema.org/draft/2020-12/vocab/core: true
  .. all the rest of the standard 2020-12 vocabularies...
```

Alternatively, perhaps we should simply forbid mixing and matching vocabularies from different specification versions (I already do this in my implementation), and we can escape all this madness.

**Reply by jdesrosiers** [reply]
@karenetheridge I can definitely see how it can be interpreted that way, but @handrews has been telling us that the intention is that the core vocabulary is what determines which JSON Schema version is to be used to evaluate the schema, not `$schema`. Prior to 2019-09 is was a JSON Schema version identifier _and_ a reference to the meta-schema. Since 2019-09, `$schema` is only intended to be a reference to the meta-schema and the JSON Schema version is identified by the core vocabulary URI in `$vocabulary`.

**Reply by jdesrosiers** [reply]
This weekend, I updated my implementation to use this algorithm. It works and the implementation is simple enough. However, I do have one concern.

My implementation is based on a core engine that implements the core processing rules of JSON Schema. Implementations of the various versions of JSON Schema configure that core and supply implementations for the keywords they support. The only problem with the change is that now my core engine needs to hardcode the core vocabulary URIs when it used to be completely version agnostic. If this stays the way it is, I will have to update my core engine every time there is a release, which is not great. Ideally we can update the way the JSON Schema version is identified (Option 2 from OP) so these URIs don't need to be hardcoded in the future.

**Reply by karenetheridge** [reply]
>  @handrews has been telling us that the intention is that the core vocabulary is what determines which JSON Schema version is to be used to evaluate the schema, not $schema

This is tricky because we don't know that we need to evaluate $vocabulary until we know we're working with a spec version that supports it, and there could be a contradiction there.  e.g. what happens in this case?

```json
{
  "$schema": "https://example.com/my-metaschema",
  ...,
}
```

```json
{
  "$id": "https://example.com/my-metaschema",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$vocabulary": {
    "https://json-schema.org/draft/2020-12/vocab/core": true
  }
}
```

I *must* look at the $schema keyword there to figure out what core keywords to evaluate and which to ignore.  Hiding the version semantics in $vocabulary is too late.  (Unless we prohibit mixing and matching $schema and $vocabulary versions, in which case we are effectively determining version semantics from $schema, not $vocabulary.)

Perhaps we are not quite talking about the same thing.  (FWIW I already hardcode the spec metaschema and vocabulary URIs and use lookup tables to map those to versions - if a draft2022-07 came out that was functionally identical to draft2020-12, I would still need to make updates in order to support it. indeed that was the first commit in my draft-next branch -- adding additional URIs to the map tables, and keeping the list of supported keywords as identical to draft2020-12.)

**Reply by awwright** [reply]
> JSON Schema version identifier

To be clear, schemas are unversioned—only JSON Schema the specification has versions.

**Reply by awwright** [reply]
> My point is that everything can be derived from $schema and then parsing the document it refers to (and so on and so on),

Here's a problem I don't think we've internalized yet: Meta-schemas are not transitive. What does that mean?

An instance is _exclusively_ described by the schema; it is not described by the meta-schema. The schema is described _exclusively_ by the meta-schema; the meta-meta schema has no say.

So if you have a schema and meta-schema

```javascript
{
  "$schema": "https://example.com/my-metaschema",
  ...,
}
```

```javascript
{
  "$id": "https://example.com/my-metaschema",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$ref": "http://json-schema.org/draft-04/schema#"
}
```

You wouldn't consult the $schema, that only tells you how to interpret "my-metaschema". You would consult the $ref, that is the schema that will "import" draft-04 semantics into the meta-schema.

**Reply by gregsdennis** [reply]
To further @awwright point above (and I _think_ this aligns with what @handrews and @jdesrosiers are saying), a meta-schema's `$vocabulary` defines what keywords can be used in the schema.  That is, whereas `$schema` only defines how to interpret the _current_ document, `$vocabulary` defines how to interpret derived documents.

This is further evidenced by the spec ([section 8.1.2](https://json-schema.org/draft/2020-12/json-schema-core.html#name-the-vocabulary-keyword)):

> The "$vocabulary" keyword is used in meta-schemas to identify the vocabularies available for use in schemas described by that meta-schema.

and

> The "$vocabulary" keyword MUST be ignored in schema documents that are not being processed as a meta-schema.

So expanding the example above:

```json
{
  "$schema": "https://example.com/my-metaschema",
  "$vocabulary": {
    ...
  }
  ...,
}
```

```json
{
  "$id": "https://example.com/my-metaschema",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$ref": "http://json-schema.org/draft-04/schema#"
}
```

The meta-schema's `$vocabulary` directly defines how to interpret what's in the schema, but the meta-schema's `$schema` (the meta-meta-schema) only defines how to interpret what's in the meta-schema and has no direct impact on the schema.

---

I _think_ my implementation does this but it also does the follow-`$schema`-until-it-finds-something-it-knows thing.  Each keyword is marked with the draft and vocabulary that define it (sometimes a keyword is defined by multiple drafts or vocabs, like `minLength`).  The first step in validation is determining which keywords to process.  This is first done by looking at which keywords are defined by the meta-schema in `$schema`.  Then, if that meta-schema also defines a `$vocabulary`, it _further_ filters keywords by those vocabs.

I believe we have a test around this where we define a custom meta-schema that explicitly omits the validation vocab, and then we have the schema attempt to use a keyword defined in it.  The test ensures that the keyword does not attempt assertion.

I'm not sure what my implementation would do with the mixing of vocabs and `$schema`.  For example, what happens here (similar to @karenetheridge's example):

```json
{
  "$id": "https://example.com/my-metaschema",
  "$schema": "https://json-schema.org/draft/2019-09/schema",
  "$vocabulary": {
    "https://json-schema.org/draft/2020-12/vocab/core": true,  // or maybe even 2019-09 for core
    "https://json-schema.org/draft/2020-12/vocab/applicator": true
  }
}
```

```json
{
  "$schema": "https://example.com/my-metaschema",
  "prefixItems": [{ "type": "string" }]
}
```

I believe it would skip `prefixItems` because it's not defined in 2019-09, so it would be eliminated in the first filtering pass.

From what I'm reading above, @handrews and @jdesrosiers are saying that `prefixItems` should be processed.

**Reply by jdesrosiers** [reply]
The meta-schema in @karenetheridge's example is a draft-07 schema and therefore `$vocabulary` can't be used to determine the JSON Schema version for the original schema. So, the JSON Schema version of the original schema is indeterminate. All of our implementations assume that the original schema is draft-07 because it's meta-schema is draft-07, which is fine, but that behavior isn't defined in the spec.

Let's use the OpenAPI dialect as a non-trivial, but not too complex example. Here's a schema using that dialect.

```json
{
  "$id": "https://example.com/my-schema",
  "$schema": "https://spec.openapis.org/oas/3.1/dialect/base",
  ...
}
```

The meta-schema is.

```json
{
  "$id": "https://spec.openapis.org/oas/3.1/dialect/base",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$vocabulary": {
    "https://json-schema.org/draft/2020-12/vocab/core": true,
    ...
    "https://spec.openapis.org/oas/3.1/vocab/base": false
  },
  ...
}
```

At this point we don't know what kind of schema this is, so we have to look at `$schema` to get the meta-meta-schema.

```json
{
  "$id": "https://json-schema.org/draft/2020-12/schema",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$vocabulary": {
    "https://json-schema.org/draft/2020-12/vocab/core": true,
    ...
  }
  ...
}
```

Since `$schema` and `$id` are the same, we can now use the induction proof I described to determine that this is a 2020-12 schema. Therefore, the `$vocabuary` keyword in `https://json-schema.org/draft/2020-12/schema` tells us that `https://spec.openapis.org/oas/3.1/dialect/base` is a 2020-12 schema. Therefore, the `$vocabulary` keyword in `https://spec.openapis.org/oas/3.1/dialect/base` tells us that `https://example.com/my-schema` is a 2020-12 schema.

I believe this is essentially the same process we were all already implementing except for how you identify the type of schema when you get to a self describing schema.

**Reply by jdesrosiers** [reply]
After writing that up yesterday, I realized that I wasn't quite doing it correctly and rewrote it. I have to take back the part where I said the implementation was simple enough. It's actually fairly complicated, but I believe it works correctly even for pathological cases and is always unambiguous. 

... unless I missed something else

**Reply by handrews** [reply]
There are two concerns:

1. It's technically possible to write a meta-schema that does not reference any standard meta-schema.  It wasn't until draft-05 that custom meta-schemas were even recommended to reference the standard meta-schemas.  We could declare such cases to have implementation-defined behavior, but they do need to be acknowledged.
2. The induction process may work (aside from the previous point), but it's convoluted.  I do not think that locking it in as our long-term strategy is a good idea at all.

**Reply by jdesrosiers** [reply]
> It's technically possible to write a meta-schema that does not reference any standard meta-schema. 

I'm not following what you mean here. The process I defined is not dependent on the standard meta-schemas. It requires a meta-schema that is _self describing_ (`$id` == `$schema`) at some point in the chain, but that doesn't need to be the official meta-schemas.

For example, the OpenAPI dialect schema could have been written like the following and works without having to get the official meta-schema involved. (I should have done it like this in first place, but I didn't have the understanding that I do now.)

```json
{
  "$id": "https://spec.openapis.org/oas/3.1/dialect/base",
  "$schema": "https://spec.openapis.org/oas/3.1/dialect/base",
  "$vocabulary": {
    "https://json-schema.org/draft/2020-12/vocab/core": true,
    ...
    "https://spec.openapis.org/oas/3.1/vocab/base": false
  },
  ...
}
```

I'm not sure that addresses your point. Sorry, if I missunderstood.

> I do not think that locking [the induction process] in as our long-term strategy is a good idea at all.

I think this misses the point. The induction proof strategy shows that we don't need to lock anything down. Technically, it's not even necessary to lock down `$schema` although I'm not sure it's necessary to go that far. We can delegate this whole process to the individual spec versions. Future releases can and should introduce something that's easier to manage than what we have now and none of it needs to be locked down in the media-type registration. The induction proof is only necessary because of the way the vocabulary system identifies the core vocabulary. We can change that in the next release and then only implementations that what to support the older version need to deal with induction proofs. Of course that would be most implementations for a while, but it puts us on a path to phasing out what we can't change.

**Reply by awwright** [reply]
> The meta-schema in @karenetheridge's example is a draft-07 schema and therefore $vocabulary can't be used to determine the JSON Schema version for the original schema. So, the JSON Schema version of the original schema is indeterminate.

One of the things I think we should be more specific about is what we mean by "JSON Schema version". The specification never uses versions; what I take "schema version" to mean is: a schema written expecting a validator that implements a particular version of the specification, for a specific result.

What this means, practically speaking, is it might be possible that some schemas are "ambiguous". They could have plausibly been written for different versions of the specification, expecting different results. _What are these cases, specifically?_

I think it's likely that when a schema version is not obvious, the behavior can be made consistent regardless.

For example with $vocabulary: Maybe we cannot figure out if a schema was written for "draft-07" or "2019-09", but this doesn't actually matter. The specification requires that $vocabulary _always_ be read... a draft-07 schema is no exception. The only reason that a validator would not use $vocabulary is if it does not implement 2019-09 or later. Likewise, the only reason a schema would embed $vocabulary is because it's using 2019-09 or later. We can make use of these facts to our advantage.

Now maybe it's the case that there's no $vocabulary and there's no known meta-schema... a certain amount of heuristics like you describe is legal, it's just not going to be interoperable (different implementations can sniff this out in ways that are incompatible with each other).

Although in some cases it won't be. If you don't read $vocabulary because the meta-meta-schema is draft-07, that might break some esoteric uses of JSON Schema.

**Reply by awwright** [reply]
> It's technically possible to write a meta-schema that does not reference any standard meta-schema.

While this may be technically legal to write, is it possible to unambiguously support this?

We can support a schema that _describes_ itself. What I don't think is possible is supporting a schema that _defines_ itself.

That is, a schema which defines itself is a tautology—a paradox. A schema can describe itself, but something else (the specification) has to be the authority that defines it.

(*) Of course, $vocabulary is what breaks this paradox, and hides definitions behind an opaque URI. $schema describes, $vocabulary defines.

**Reply by gregsdennis** [reply]
> What this means, practically speaking, is it might be possible that some schemas are "ambiguous". They could have plausibly been written for different versions of the specification, expecting different results. _What are these cases, specifically?_

A simple example of such a case could be a schema that (perhaps erroneously) includes a `$ref` with siblings.

```json
{
  "$ref": "just-an-integer.json",
  "minimum": 5
}
```

Without knowing which version of the specification to use when interpreting this, we may or may not apply `minimum`, and the spec currently lists this as "implementation defined."

All of these cases would have to involve keywords that didn't change syntax but did change semantics, while possibly also including keywords that didn't change at all.  We don't often change keyword semantics without also changing syntax, but it has happened.

- `items` changed to disallow the array form (creating `prefixItems` for that functionality).  This is detectable because if we get an array-form `items`, we know it must be from a draft prior to `prefixItems` being introduced.  At a minimum, it reduces the gamut of spec versions that _could_ apply to ones that all have the same semantics for this keyword, so it doesn't matter which one we pick.
- `$ref` changed to allow siblings to be processed.  Because this change has no syntactic difference, we don't get the same gamut reduction, and we still end up with a potential set of applicable spec versions that prescribe different semantics.

**Reply by jdesrosiers** [reply]
> The specification requires that $vocabulary always be read... a draft-07 schema is no exception.

You keep saying this, but it doesn't seem like anyone agrees with your interpretation. If anyone disagrees, please speak up, but otherwise I'd like to stop rehashing this subject. Respecting a keyword that didn't exist in the version of the specification that the schema declares it was written for is not correct.

I'm sorry to be so direct about this, but we aren't going to get anywhere if we keep talking past each other. We need to decide on an interpretation and more on.

**Reply by jdesrosiers** [reply]
> Now maybe it's the case that there's no $vocabulary and there's no known meta-schema... a certain amount of heuristics like you describe is legal, it's just not going to be interoperable (different implementations can sniff this out in ways that are incompatible with each other).

The process I've described here is not a heuristic. It shows how to unambiguously figure out which specification version should be used based on the rules defined in each specification version. There is no guessing. There is no out of band information.

Of course it is possible to encounter a schema where it can't be determined which version of the spec was intended and it is allowed to use heuristics to guess how to interpret the schema in that case, but that's not what this thread is about. Anything other than refusing to process the schema in that case risks getting unexpected and unintended results and should be discouraged in my opinion.

**Reply by awwright** [reply]
> `$ref` with siblings

I think this is sort of a special case, and sort of the same case as $vocabulary.

Nobody was writing $vocabulary in their schemas, expecting it to be ignored. So we can safely rely on the presence of $vocabulary to indicate something about the author's expectations. Likewise, nobody was writing validation keywords next to $ref and expecting them to be ignored.

To the contrary, people were writing keywords next to $ref, they were expecting them to be _processed_. This was limited to annotation keywords like "title". But it was enough for us to change the behavior to match this _lawlessness_ (/s).

---

But for the sake of argument, let's assume that ignoring siblings to `$ref` is really important to someone. Is there anything we can do about this? In this case, it genuinely would be ambiguous behavior, and all we can do is (1) leave it implementation-defined, or (2) pick a winner.

**Reply by awwright** [reply]
@jdesrosiers I call it a "heuristic" because we don't have a written standard. There is no guidance on how one validator may incorporate historical specifications into a validator. Different implementations may legitimately do it differently, and there's nothing to say one is correct and the other is wrong.

Example: If I implement _only_ 2020-12, I will read in $vocabulary every time, even if the $schema says draft-07. Is that wrong? You do it differently? Who's right?

When you're varying from the behavior specified in JSON Schema Core, you're necessarily implementing a heuristic that decides when to do so. I'm not even saying that's wrong! I'm saying we should be clear about how this is possible.

---

And the reason I bring it up here is to point out that that, best case, the specification text is contradictory. There may be someone in some esoteric use case who writes draft-07 meta-schemas with $vocabulary and found that it worked, and expects that to work. Who "wins"?

I've described what kinds of backwards compatibility the specification does and does not support—but I think we agree this is insufficient and the point of this discussion is how do we change it. This is why I'm asking for some concrete examples.

**Reply by awwright** [reply]
> It shows how to unambiguously figure out which specification version should be used based on the rules defined in each specification version. There is no guessing. There is no out of band information.

In [my comment earlier](https://github.com/orgs/json-schema-org/discussions/198#discussioncomment-3133241), I introduced an example that is ambiguous. I was hoping you would address this?

Here is the complete example of the meta-schema:

```javascript
{
"$id": "http://example.com/ambiguous-meta-schema"
"$schema": "https://json-schema.org/draft-07/schema#",
"$vocabulary": { "https://json-schema.org/draft/2020-12/vocab/core": true },
"$ref": "https://json-schema.org/draft/2019-09/schema"
}
```

Note how the behavior will be different depending on if the validator understands the draft-07 meta-schema or not.

Here is a slight modification of your algorithm that moves the "$schema" test lower down to resolve this ambiguity:

```
IF the meta-schema has $vocabulary["https://json-schema.org/draft/2019-09/vocab/core"] == true
  schemaType = "https://json-schema.org/draft/2019-09/vocab/core"
ELSE IF the meta-schema has $vocabulary["https://json-schema.org/draft/2020-12/vocab/core"] == true
  schemaType = "https://json-schema.org/draft/2020-12/vocab/core"
ELSE IF $schema is a known dialect
  schemaType = $schema
ELSE
  I don't know how to process this schema. I might guess, use a default, or throw an error.
```

While this fixes that ambiguity, there's still other ambiguities (what if I remove $vocabulary... does draft-07 win or does 2019-09 win?). This is particularly interesting because 2019-09 says that $vocabulary does not cross $ref boundaries. I'm still researching this.

**Reply by handrews** [reply]
@jdesrosiers 

> The process I defined is not dependent on the standard meta-schemas. It requires a meta-schema that is self describing (`$id` == `$schema`) at some point in the chain, but that doesn't need to be the official meta-schemas.

Ah, right- you are correct.  However, a self-describing meta-schema that is also non-standard (and does not use `$vocabulary`) is an ambiguous case.

I don't agree with the rest of your response, but I don't have the spoons to unravel it right now.  I know you're in a rush to do this, but I just cannot keep up with the pace of this conversation right now.

**Reply by jdesrosiers** [reply]
@awwright 
> I call it a "heuristic" because we don't have a written standard. There is no guidance on how one validator may incorporate historical specifications into a validator. 

Ok. I see how you got there. However, I don't agree with your conclusion. Even though no individual spec tells you what to do, you can apply the rules from each spec version and get an unambiguous result. There are cases where no spec versions will claim the schema. In those cases you can use a heuristic or refuse the schema, but those aren't the cases I'm talking about in this thread. I'm convinced that it's not possible to have two different spec versions lay claim to the same schema.

You're example is the same as @karenetheridge's that I addressed in https://github.com/orgs/json-schema-org/discussions/198#discussioncomment-3124915. I'll try to explain in more detail.

```json
{
  "$id": "http://example.com/ambiguous-meta-schema"
  "$schema": "https://json-schema.org/draft-07/schema#",
  "$vocabulary": { "https://json-schema.org/draft/2020-12/vocab/core": true },
  "$ref": "https://json-schema.org/draft/2019-09/schema"
}
```

If we apply the draft-07 rules, we see `"$schema": "https://json-schema.org/draft-07/schema#"` and we know this is a draft-07 schema.

If we apply the 2020-12 rules, we have to look at the `https://json-schema.org/draft-07/schema` schema in order to determine which spec version to use for this schema. That schema does not have a `$vocabulary` keyword and therefore, the spec version to use for the original schema is unknown.

Therefore, the schema is unambiguously a draft-07 schema.

What if someone uses this schema as a meta-schema?

```json
{
  "$id": "https://example.com/my-schema",
  "$schema": "http://example.com/ambiguous-meta-schema",
  ...
}
```

`http://example.com/ambiguous-meta-schema` is not associated with a known spec version, so no spec version draft-07 or lower is going to claim this schema.

The 2019-09 or 2020-12 specs say that we should look at the meta-schema to determine which spec version to use for this schema. We have determined that the meta-schema is a draft-07 schema, so even tho it has a `$vocabulary` keyword that declares the 2020-12 core vocabulary, that declaration is ignored because `$vocabulary` is not a draft-07 keyword. Therefore, the spec version to use to interpret the schema is unknown.

None of the spec versions claim this schema so at this point you can use a heuristic to guess or refuse the schema.

**Reply by awwright** [reply]
@jdesrosiers Ok. I _don't_ think it's the case that you can unambiguously combine the drafts together like you're suggesting (that there's going to have to be some cases where someone's use case is going to break). This may be because of the use cases I'm trying to support and I'm working on putting together some comprehensive examples.

**Reply by awwright** [reply]
@jdesrosiers Could you speak specifically to my example where I implement only 2020-12?

The issue in this case isn't backwards compatibility or properly handing any current schema, I doubt anyone in the wild has written something like that.

But it seems that later on, some implementations will recognize draft-07, and other implementations will not; so even though both may be up-to-date, their behavior will still be different. So someone might be able to write a schema that only works on one of those implementations, and exploit this.

**Reply by jdesrosiers** [reply]
> Could you speak specifically to my example where I implement only 2020-12?

This is not a special case. It works the same. This is a 2020-12 implementation, so it needs to look at the meta-schema to know what spec version to use. It finds the standard draft-07 meta-schema which is something the implementation doesn't support. Therefore, it needs to guess how to interpret the original schema or refuse to process it. If it guesses, there could be interoperability issues, which is why the correct behavior IMO is to refuse the schema.

**Comment by handrews** [comment]
Because of this:

> The approach we are taking is to standardize only dialect identification in the media type registration 
...
> However, this doesn't technically work for 2019-09 and 2020-12 because it's not $schema that determines how a schema is processed, it's $vocabulary. 

Can we talk about standardizing _processing rules identification_ instead of dialect identification?  The dialect is actually irrelevant except when it's the only available identifier for the processing rules.  And in that case it's serving two roles rather than really being about dialect.

(I will come back and add an example about the `$schema`-of-meta-schema problem, and also why I think option #2 is the safest option as soon as I get some time.  Which may take a few days.)

**Reply by awwright** [reply]
> The dialect is actually irrelevant except when it's the only available identifier for the processing rules.

This is an important insight. It doesn't seem possible to "standardize only dialect identification".

**Reply by jdesrosiers** [reply]
I don't think it's possible to standardize "rules identification" in a way that works for past and future releases without adding a keyword that retroactively applies to past releases and pins us to a solution for future releases. Both of those things are things I'd like to avoid.

Different specifications have different methods of rules identification and future specifications may have different methods as well. That's why I think it makes sense to go one step higher and just standardize dialect identification. That allows implementations to determine rules identification based on which specifications they support and doesn't constrain our future choices either. (See https://github.com/orgs/json-schema-org/discussions/198#discussioncomment-3103803)

**Reply by awwright** [reply]
I want to make sure I understand this correctly so I asked a question below first... however I was going to suggest something like the algorithm that you posted there. As long as there is some way to determine "the core semantics have changed in an incompatible way" it doesn't have to be $schema. It can be e.g. the lack of a known core vocabulary. There's a few tweaks that I'm ironing out, but this is a much better path.

**Comment by awwright** [comment]
> What we define should work for all past a future releases of JSON Schema and their existing implementations

What does this mean in technical terms? Can we come up with some examples of things that we would like to enable, that would not be possible just by referencing 2020-12?

**Reply by jdesrosiers** [reply]
Can you be more specific. Is there a specific part of 2020-12 you want to reference? Or do you mean referencing the whole thing?

I don't see how pinning any part of the media-type registration to a specific version could possibly be feasible. It excludes support for past and future releases. I know you see the current release as inclusive of past releases, but that's not how everyone else interprets it. Even if that was the case, we would still have to update the registration every time we do a new release to change it point to the new release. Then there's third party dialects like OpenAPI and MongoDB that wouldn't fit in that model at all.

**Reply by awwright** [reply]
Here's an example I'm looking for:

JSON Schema standardizes a new keyword that lets you validate a string against a deterministic context-free grammar (i.e. regular expression with recursion), and I want add such a test to my JSON Schema. Correct validation is critical to the security of the application, so I want to ensure that all validators validate the new keyword—they must not say that bad instances are "valid", but they may fail validation or raise an error.

Can we ensure this, even for validators that haven't been updated to the latest spec? (Assume that only validators compliant with the algorithm we're going to define here are used. e.g. Suppose there's a `validator.enableStrict()` call that will guarantee this behavior; so that validators without this behavior will produce a "call to undefined value" error.)

**Reply by awwright** [reply]
Maybe there's actually a way to guarantee this if I know the validator uses 2020-12. But suppose a validator in the far future comes along and doesn't implement 2020-12 at all. How can I guarantee that the validator isn't just going to make an educated guess?

**Reply by awwright** [reply]
So with regard to my original question: Let's come up with more specific examples that we can judge potential solutions by.

With respect to my example: If security like this is important, then at some point we're going to have to come up with a forward compatibility scheme and commit to it. That scheme can be something like "Dereference the meta-schema and check for the existence of a known core URI. If none is found, error."

Note that this does _not_ commit meta-schemas to using $vocabulary forever. This is because we can add exceptions: "Dereference the meta-schema; if there's no $vocabulary _and_ no `$core` keyword, then error." (`$core` being a hypothetical keyword that replaces `$vocabulary`).

**Reply by jdesrosiers** [reply]
> Can we ensure this, even for validators that haven't been updated to the latest spec?

In the normal case, the dialect / processing rules of the schema should be identified somehow. In this case, it's probably part of a vocabulary with the "must understand" flag set to `true` and the schema will be rejected because it doesn't understand the new keyword.

In the unusual case that a dialect can not be determined, technically all bets are off and the client can choose to do anything including processing the schema and ignoring the new keyword. However, the more likely behavior is that clients will either be very clear about what they default to if they don't know something or better yet, just refuse to process the schema.

We could define the registration to require that a dialect is known, but I don't think that's technically compatible with older versions of the spec.

**Reply by jdesrosiers** [reply]
I don't think we can define backwards/forwards compatibility into the media type registration because past releases are not backwards/forwards compatible. That kind of thing needs to be defined in future specifications. It's an important thing that we need to discuss and solve for in the next release, but I don't think it's relevant to this discussion.

**Reply by awwright** [reply]
@jdesrosiers So what are some of your use cases? It seems to me this should enable us to do something specific, otherwise we would just write "Do something JSON Schema-like, dunno how, use your best judgement." So what are those things?

**Reply by jdesrosiers** [reply]
I want to be able to annotate JSON responses in an API context using a Link header with a "describedby" relation.

I want a validator implementations that enable fetching schemas over HTTP to use proper web conventions. I want to reject a document that doesn't have a proper content type as not a JSON Schema document even if it happens to look like a JSON Schema document. Optionally, if the schema doesn't declare what dialect it uses, I want to be able to get that information from the media type, but I'd prefer that information is in the schema. For security reasons, I want to be able to refuse to process any schema that doesn't somehow declare its dialect. However, clients also need to be allowed to assume a dialect in certain cases like older OpenAPI specifications that don't have a dialect id to declare. This is necessary for backwards compatibility, but should be discouraged.

I want to be able to content negotiated what dialect of a schema I can handle when fetching a schema over HTTP.

There's at least one more use case that the media type is designed to support, but those are the ones I personally use.

**Comment by OR13** [comment]
I'm supportive of getting these registered... let me know how I can help.
